Duplicati/UnitTest/SymLinkTests.cs

```diff
@@ -0,0 +1,157 @@
+                    Assert.AreEqual(0, restoreResults.Errors.Count());
+                    Assert.AreEqual(0, restoreResults.Warnings.Count());
+
+                    // Verify that symlink policy was followed
+                    var restoreSymlinkDir = systemIO.PathCombine(this.RESTOREFOLDER, symlinkDirName);
+                    switch (symlinkPolicy)
+                    {
+                        case Options.SymlinkStrategy.Store:
+                            // Restore should contain an actual symlink to the original target
+                            Assert.That(systemIO.IsSymlink(restoreSymlinkDir), Is.True);
+                            var restoredSymlinkFullPath = systemIO.PathGetFullPath(systemIO.GetSymlinkTarget(restoreSymlinkDir));
+                            var symlinkTargetFullPath = systemIO.PathGetFullPath(targetDir);
+                            Assert.That(restoredSymlinkFullPath, Is.EqualTo(symlinkTargetFullPath));
+                            break;
+                        case Options.SymlinkStrategy.Follow:
+                            // Restore should contain a regular directory with copies of the files in the symlink target
+                            Assert.That(systemIO.IsSymlink(restoreSymlinkDir), Is.False);
+                            AssertDirectoryTreesAreEquivalent(targetDir, restoreSymlinkDir, $"Symlink policy: {Options.SymlinkStrategy.Store}");
+                            break;
+                        case Options.SymlinkStrategy.Ignore:
+                            // Restore should not contain the symlink at all
+                            Assert.That(systemIO.DirectoryExists(restoreSymlinkDir), Is.False);
+                            break;
+                    }
+                    // Prepare for a fresh backup and restore cycle
+                    foreach (var dir in systemIO.EnumerateDirectories(this.RESTOREFOLDER))
+                    {
+                        systemIO.DirectoryDelete(dir, true);
+                    }
+                    foreach (var file in systemIO.EnumerateFiles(this.RESTOREFOLDER))
+                    {
+                        systemIO.FileDelete(file);
+                    }
+                    c.DeleteAllRemoteFiles();
+                    if (systemIO.FileExists(this.DBFILE))
+                    {
+                        systemIO.FileDelete(this.DBFILE);
+                    }
+                    if (systemIO.FileExists($"{this.DBFILE}-journal"))
+                    {
+                        systemIO.FileDelete($"{this.DBFILE}-journal");
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Asserts that the two directories are equivalent; i.e., they they contain the same subdirectories and files, recursively.
+        /// </summary>
+        private static void AssertDirectoryTreesAreEquivalent(string expectedDir, string actualDir, string contextMessage)

```

Comment:
The `TestUtils.VerifyDir` method seems to have a similar intent.  I do like how this recursive implementation is simpler.  Is there any reason to prefer one over the other?

Reply:
`TestUtils.VerifyDir` throws an exception for some problems (e.g., for missing folders or files), but only writes log messages for others (e.g., if the contents of two files are different).  On the other hand, `AssertDirectoryTreesAreEquivalent` uses the NUnit Assert mechanism for any problems it encounters, including if the contents of two files are different.  Also, `TestUtils.VerifyDir` (via `TestUtils.CompareFiles`) verifies some file metadata, while `AssertDirectoryTreesAreEquivalent` and `AssertFilesAreEqual` don't, though that would be easy to add.

I generally like using the NUnit Assert messages to capture context that can make the source of a failure more aparent.  It keeps the failure reasons localized in the NUnit failure report.  For example, here's what NUnit reports when `AssertDirectoryTreesAreEquivalent` fails due to file contents not being equal:
```
1) Failed : Duplicati.UnitTest.SymLinkTests.SymLinkPolicy
  Symlink policy: Store, file contents mismatch at position 0 for C:\D\td\backup-data\target\b.txt and C:\D\td\restored\symlink\b.txt
  Expected: 98
  But was:  65
```

I'll take this in any direction you'd like.  Just let me know how you'd like to proceed.

Reply:
I took a look at `TestUtils.VerifyDir`, and you're correct.  This implementation is much better.  Can we:
1. Move `AssertDirectoryTreesAreEquivalent` to `TestUtils`
2. Replace usages of `VerifyDir` with `AssertDirectoryTreesAreEquivalent`
3. Remove `VerifyDir`.
4. Move `AssertFilesAreEqual` to `TestUtils`
5. Add the ability for `AssertFilesAreEqual` to compare metadata
6. Replace usages of `CompareFiles` with `AssertFilesAreEqual`
7. Remove `CompareFiles`

Reply:
Will do, thanks!

Reply:
I've pushed a commit that includes these changes.